JavaScript/JavaScriptSDK/Context/Sample.ts

```diff
@@ -2,6 +2,57 @@
     "use strict";
 
     export class Sample {
-        public sampleRate: string;
+        public sampleRate: number;
+
+        public INT_MAX_VALUE: number = 2147483647;
+
+        constructor(sampleRate: number) {
+            if (sampleRate > 100 || sampleRate < 0) {
+                _InternalLogging.throwInternalUserActionable(LoggingSeverity.WARNING, "Sampling rate is out of range (0..100): '" + sampleRate
+                    + "'. Sampling will be disabled, you may be sending too much data which may affect your AI service level.");
+                this.sampleRate = 100;
+            }
+
+            this.sampleRate = sampleRate;
+        }
+
+        /**
+        * Determines if an envelope is sampled in (i.e. will be sent) or not (i.e. will be dropped).
+        */
+        public isSampledIn(envelope: Telemetry.Common.Envelope): boolean {            
+            if (this.sampleRate == 100) return true;
+                        
+            // TODO: extract to SamplingScoreGenerator
+            var score = this.getScore(envelope);
+            
+            return score * 100 < this.sampleRate;
+        }
+
+        public getScore(envelope: Telemetry.Common.Envelope): number {            
+            var tagKeys: AI.ContextTagKeys = new AI.ContextTagKeys();
+            var score: number = 0;
+            if (envelope.tags[tagKeys.userId]) {
+                score = Sample.getSamplingHashCode(envelope.tags[tagKeys.userId]) / this.INT_MAX_VALUE;
+            } else if (envelope.tags[tagKeys.operationId]) {
+                score = Sample.getSamplingHashCode(envelope.tags[tagKeys.operationId]) / this.INT_MAX_VALUE;
+            } else {
+                score = Math.random()
+            }
+
+            return score;
+        }
+
+        public static getSamplingHashCode(input: string): number {
+            if (input == "") { return 0; }
+
+            var hash: number = 5381;
```

Comment:
Alex, Bing it! :) http://stackoverflow.com/questions/1579721/why-are-5381-and-33-so-important-in-the-djb2-algorithm